src/Kernel/Magic/QirMagic.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Jupyter.Core;
+using Microsoft.Quantum.IQSharp.AzureClient;
+using Microsoft.Quantum.IQSharp.Common;
+using Microsoft.Quantum.IQSharp.Jupyter;
+using Microsoft.Quantum.QsCompiler;
+
+namespace Microsoft.Quantum.IQSharp.Kernel
+{
+    /// <summary>
+    ///     A magic command that can be used to generate QIR from a given
+    ///     operation as an entry point.
+    /// </summary>
+    public class QirMagic : AbstractMagic
+    {
+        private const string ParameterNameOperationName = "__operationName__";
+        private const string ParameterNameOutputPath = "output";
+
+        public QirMagic(ISymbolResolver resolver, IEntryPointGenerator entryPointGenerator, ILogger<SimulateMagic> logger) : base(
+            "qir",
+            new Microsoft.Jupyter.Core.Documentation
+            {
+                Summary = "TODO.",
+                Description = @"
+                    TODO
+
+                    #### Required parameters
+
+                    - Q# operation or function name. This must be the first parameter, and must be a valid Q# operation
+                    or function name that has been defined either in the notebook or in a Q# file in the same folder.
+                ".Dedent(),
+                Examples = new []
+                {
+                    @"
+                        TODO
+                    ".Dedent()
+                }
+            }, logger)
+        {
+            this.EntryPointGenerator = entryPointGenerator;
+            this.Logger = logger;
+        }
+
+        private ILogger? Logger { get; }
+
+        // TODO: EntryPointGenerator might should move out of the Azure Client
+        //       project.
+        public IEntryPointGenerator EntryPointGenerator { get; }
```

Comment:
Is this TODO being tracked as an issue? It seems like a possibly large change.

Reply:
The EntryPointGenerator is supposed to be moved out of the Azure Client project? I'm not sure why that would be desirable. I will make a GitHub issue for it anyway.

Reply:
Issue: #610

Reply:
Thanks Scott!

Reply:
> I'm not sure why that would be desirable.

For the reasoning there, the Azure client should be specific to what's needed to submit to Azure, but we use entry point generation in local QIR workflows as well.